### ISSUE TITLE
DMREMB is not changed after the call to CQR1676::decode(DMREMB)

### ISSUE THREAD

**Issue by shawnchain** [g4klx/MMDVMHost#541]
https://github.com/g4klx/MMDVMHost/blob/6713e0f52b0e74fcf2e72cb91cdcdb7328e6c026/DMREMB.cpp#L47

It seems that the DMREMB[2U] has never been changed inside the method  `CQR1676::decode(DMREMB)` which means QR1676 decoding is not applied to the received 2 bytes EMB data ?

**Comment by g4klx** [comment]
The EMB is only 16-bits in length, so only two bytes of the data array are used. I would not expect DMREMB[2U] to be accessed.

**Comment by shawnchain** [comment]
I would like to quote the code snippet here:

```
void CDMREMB::putData(const unsigned char* data)
{
	assert(data != NULL);

	unsigned char DMREMB[2U];
	DMREMB[0U]  = (data[13U] << 4) & 0xF0U;
	DMREMB[0U] |= (data[14U] >> 4) & 0x0FU;
	DMREMB[1U]  = (data[18U] << 4) & 0xF0U;
	DMREMB[1U] |= (data[19U] >> 4) & 0x0FU;

	CQR1676::decode(DMREMB);   <---- *** DMREMB[0] has never been modified.
 instead, a char is returned but ignored, I guess this returned char it's the value
 to be used later to populate CC/PI/LCSS. Correct me if I'm wrong.

	m_colorCode = (DMREMB[0U] >> 4) & 0x0FU;
	m_PI        = (DMREMB[0U] & 0x08U) == 0x08U;
	m_LCSS      = (DMREMB[0U] >> 1) & 0x03U;
}
```

**Comment by g4klx** [comment]
No, you’re right. That’s a very old bug! When there are no bit errors it will work correctly of course.
Can you do a pull request to fix it?
It needs testing although I probably tested QR1676 when I wrote it.


Sent from Yahoo Mail for iPhone


On Friday, December 6, 2019, 14:29, Shawn <notifications@github.com> wrote:


I would like to quote the code snippet here:
void CDMREMB::putData(const unsigned char* data)
{
	assert(data != NULL);

	unsigned char DMREMB[2U];
	DMREMB[0U]  = (data[13U] << 4) & 0xF0U;
	DMREMB[0U] |= (data[14U] >> 4) & 0x0FU;
	DMREMB[1U]  = (data[18U] << 4) & 0xF0U;
	DMREMB[1U] |= (data[19U] >> 4) & 0x0FU;

	CQR1676::decode(DMREMB);   <---- *** DMREMB[0] has never been modified.
 instead, a char is returned but ignored, I guess this returned char it's the value
 to be used later to populate CC/PI/LCSS. Correct me if I'm wrong.

	m_colorCode = (DMREMB[0U] >> 4) & 0x0FU;
	m_PI        = (DMREMB[0U] & 0x08U) == 0x08U;
	m_LCSS      = (DMREMB[0U] >> 1) & 0x03U;
}

**Comment by shawnchain** [comment]
Fixed in 31408d4, ab226ce
